### PR TITLE
Koji: Retrieve the Manifests for a compose job

### LINKED ISSUE #996: kojiapi: ablity to get the manifests for a compose job
If a compose fails, It probably would be handy to get the complete manifests for all the image builds and attach that as a log to the (failed) koji task. Could be combined with the full log API or a separate call.

### PR DESCRIPTION
**Pull request by achilleas-k**
Added an API endpoint that returns the manifest for a given osbuild-koji job.

The Koji API function assumes that the second dependency of the incoming job ID (which should match a Finalize Job) is the first (and perhaps only) Build Job. It retrieves the arguments of the Build Job to extract the Manifest.

The tests cover basic (non-failing) cases for now. More notable is that in all test cases the manifest is empty, which isn't very interesting.

Closes #996.

### CODE REVIEW

**Review by gicmo — COMMENTED** [review]
Very cool! Looks very good. One thing, since the manifest is different for each architecture, I think we want the manifest for all the builds; vermuch as with status and logs. :D

**Comment by achilleas-k** [comment]
> Very cool! Looks very good. One thing, since the manifest is different for each architecture, I think we want the manifest for all the builds; vermuch as with status and logs. :D

Thanks.  Should have realised that the manifests are different after being submitted (even though the same request is sent) since different archs have different hashes and maybe even packages after depsolve.

File: internal/kojiapi/api/openapi.yml [hunk 1]
```diff
@@ -160,6 +192,10 @@ components:
         koji_import_logs: {}
         image_logs:
           type: array
+    Manifest:
```
Comment (ondrejbudai) [inline internal/kojiapi/api/openapi.yml:195]: Will this definition still be valid after the DAG changes in osbuild, @gicmo? Maybe, we should just leave `Manifest` as a generic object here and add a note that it's a valid osbuild manifest with a link to https://www.osbuild.org/man/osbuild-manifest.5?

File: internal/kojiapi/server.go [hunk 2]
```diff
@@ -384,6 +384,43 @@ func (h *apiHandlers) GetComposeIdLogs(ctx echo.Context, idstr string) error {
...
+	}
+
+	var finalizeResult worker.KojiFinalizeJobResult
+	_, deps, err := h.server.workers.JobStatus(id, &finalizeResult)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Job %s not found: %s", idstr, err))
+	}
+
+	if len(deps) > 1 {
+		// Get the first build job ID (second dependency) to get its manifest
```
Comment (ondrejbudai) [inline internal/kojiapi/server.go:401]: Hmm, different architectures have different manifests so we probably need to return all of them.

File: internal/kojiapi/server.go [hunk 3]
```diff
@@ -155,7 +155,7 @@ func (h *apiHandlers) PostCompose(ctx echo.Context) error {
 		Release: request.Release,
 	})
 	if err != nil {
-		// This is a programming errror.
+		// This is a programming error.
```
Comment (ondrejbudai) [inline internal/kojiapi/server.go:158]: :heart:

File: internal/worker/server_test.go [hunk 4]
```diff
@@ -142,6 +142,43 @@ func TestUpdate(t *testing.T) {
...
+	jobId, err := server.EnqueueOSBuild(arch.Name(), &job)
+	require.NoError(t, err)
+
+	_, _, _, args, _, err := server.RequestJob(context.Background(), arch.Name(), []string{"osbuild"})
+	require.NoError(t, err)
+	require.NotNil(t, args)
+
+	var jobArgs worker.OSBuildJob
+	rawArgs, err := server.JobArgs(jobId, &jobArgs)
+	if err != nil {
```
Comment (ondrejbudai) [inline internal/worker/server_test.go:175]: I suggest using `require.NoError()` (there are in total 4 cases of `if err != nil`).

**Review by ondrejbudai — CHANGES_REQUESTED** [review]
Yeah, I have the same concern as Christian. Otherwise, this looks awesome! :heart_eyes: I have just minor nits inline.

File: internal/kojiapi/api/openapi.yml (hunk 1 above)
Comment (larskarlitski) [reply internal/kojiapi/api/openapi.yml:195]: Yeah, I'd leave it opaque.

**Review by larskarlitski — COMMENTED** [review]
Really nice work! Nothing to add beyond the other comments.

As for testing, it might make sense to exercise this route once in `test/api/koji.sh`.

File: internal/kojiapi/api/openapi.yml (hunk 1 above)
Comment (larskarlitski) [reply internal/kojiapi/api/openapi.yml:195]: Yeah, I'd leave it opaque.

**Comment by achilleas-k** [comment]
Addressed all your comments.  It's now also rebased on #1158 since that changes the `fsjobqueue` testing which this PR adds to.

File: internal/kojiapi/api/openapi.yml [hunk 5]
```diff
@@ -87,6 +87,36 @@ paths:
             text/plain:
               schema:
                 type: string
+  '/compose/{id}/manifest':
```
Comment (larskarlitski) [inline internal/kojiapi/api/openapi.yml:90]: This might be nitpicky, but should this not be `manifests` now? At least it should be documented as such, I think.

Which order are the manifests returned in?

Comment (achilleas-k) [reply internal/kojiapi/api/openapi.yml:90]: You're right.  I updated the description to mention it returns multiples, but the path and summary should be pluralised too.

The order of the manifests _should_ be the same order as when submitted.  The order should be stable for the `fsjobqueue` since dependencies are defined in an array in memory and on disk.  This will likely become an issue with the `dbjobqueue` ([see relevant comment](https://github.com/osbuild/osbuild-composer/discussions/1149#discussioncomment-275388)).

Logs use the same logic (and will have the same issue).

Comment (larskarlitski) [reply internal/kojiapi/api/openapi.yml:90]: Yeah I was wondering about how we handle logs. Thanks!

File: internal/kojiapi/api/openapi.yml (hunk 1 above)
Comment (gicmo) [reply internal/kojiapi/api/openapi.yml]: This has already been addressed, but just for the record: might indeed change.

File: internal/worker/server_test.go [hunk 6]
```diff
@@ -172,6 +172,39 @@ func TestUpdate(t *testing.T) {
...
+	imageType, err := arch.GetImageType("qcow2")
+	require.NoError(t, err)
+	manifest, err := imageType.Manifest(nil, distro.ImageOptions{Size: imageType.Size(0)}, nil, nil, nil, 0)
+	require.NoError(t, err)
+
+	tempdir, err := ioutil.TempDir("", "worker-tests-")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempdir)
+	server := newTestServer(t, tempdir)
+	server.Handler()
```
Comment (ondrejbudai) [inline internal/worker/server_test.go:188]: This is a no-op. Maybe we want to actually use `test.TestRoute` to test the also that the API works as expected?

Comment (achilleas-k) [reply internal/worker/server_test.go:188]: Thanks for catching that.  Probably missed that it's not necessary when copying from other tests.

There's no API endpoint for getting args from the **worker**.  It just grabs it directly from the `fsJobQueue`.  The `newTestServer()` is used to initialise the job queue to test the worker methods that wrap the new `JobQueue` methods.

**Review by teg — APPROVED** [review]
Nice work!

File: internal/worker/server_test.go (hunk 6 above)
Comment (ondrejbudai) [reply internal/worker/server_test.go]: Right, sorry!

**Review by ondrejbudai — APPROVED** [review]
Awesome!
